sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpEventHubProducer.cs

```diff
@@ -0,0 +1,417 @@
+                throw;
+            }
+            finally
+            {
+                EventHubsEventSource.Log.ClientCloseComplete(clientType, EventHubName, clientId);
+            }
+        }
+
+        /// <summary>
+        ///   Sends an AMQP message that contains a batch of to the associated Event Hub. If the size of events exceed the
+        ///   maximum size of a single batch, an exception will be triggered and the send will fail.
+        /// </summary>
+        ///
+        /// <param name="messageFactory">A factory which can be used to produce an AMQP message containing the batch of events to be sent.</param>
+        /// <param name="partitionKey">The hashing key to use for influencing the partition to which events should be routed.</param>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        protected virtual async Task SendAsync(Func<AmqpMessage> messageFactory,
+                                               string partitionKey,
+                                               CancellationToken cancellationToken)
+        {
+            var failedAttemptCount = 0;
+            var logPartition = PartitionId ?? partitionKey;
+            var retryDelay = default(TimeSpan?);
+            var messageHash = default(string);
+            var stopWatch = Stopwatch.StartNew();
+
+            SendingAmqpLink link;
+
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    try
+                    {
+                        using AmqpMessage batchMessage = messageFactory();
+                        messageHash = batchMessage.GetHashCode().ToString();
+
+                        EventHubsEventSource.Log.EventPublishStart(EventHubName, logPartition, messageHash);
+
+                        link = await SendLink.GetOrCreateAsync(UseMinimum(ConnectionScope.SessionTimeout, _tryTimeout)).ConfigureAwait(false);
+                        cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                        // Validate that the batch of messages is not too large to send.  This is done after the link is created to ensure
+                        // that the maximum message size is known, as it is dictated by the service using the link.
+
+                        if (batchMessage.SerializedMessageSize > MaximumMessageSize)
+                        {
+                            throw new MessageSizeExceededException(EventHubName, string.Format(Resources.MessageSizeExceeded, messageHash, batchMessage.SerializedMessageSize, MaximumMessageSize));
+                        }
+
+                        // Attempt to send the message batch.
+
+                        var deliveryTag = new ArraySegment<byte>(BitConverter.GetBytes(Interlocked.Increment(ref _deliveryCount)));
```

Comment:
I don't have much knowledge about real world scenarios, but what are the chances of a long running producer getting a `_deliveryCount` overflow?

Reply:
Potentially, yes.   It would overflow and flip to the negative minimum value.    That said, I have no understanding of how the service makes use of this at all.   I mimicked the code from track one without adjustments here.

I would speculate that it is probably just intended to be a unique value per request to help the service differentiate and that so long as two requests in a reasonable interval don't share a delivery identifier, the actual value isn't meaningful.... but, to be safe.....